model/riscv_sys_regs.sail

```diff
@@ -156,21 +156,28 @@ val sys_enable_next = {c: "sys_enable_next", ocaml: "Platform.enable_next", _: "
    unsetting C. If it returns true the write will have no effect. */
 val ext_veto_disable_C : unit -> bool effect {rreg}
 
-/* We currently only support dynamic changes for the C extension. */
 function legalize_misa(m : Misa, v : xlenbits) -> Misa = {
   if   sys_enable_writable_misa ()
-  then { /* Handle modifications to C. */
-         let  v = Mk_Misa(v);
-         /* Suppress changing C if nextPC would become misaligned or an extension vetoes or C was disabled at boot (i.e. not supported). */
-         let m =
-           if   (v.C() == 0b0 & (nextPC[1] == bitone | ext_veto_disable_C())) | not(sys_enable_rvc())
-           then m
-           else update_C(m, v.C());
-         /* Handle updates for F/D. */
-         if   not(sys_enable_fdext()) | (v.D() == 0b1 & v.F() == 0b0)
-         then m
-         else update_D(update_F(m, v.F()), v.D())
-       }
+  then {
+    let v = Mk_Misa(v);
+    /* Suppress disabling C & keep misa unchanged if nextPC would become misaligned or an extension vetoes */
+    if   v.C() == 0b0 & (nextPC[1] == bitone | ext_veto_disable_C())
+    then m
+    else {
+      /* Suppress updates to C if C is read-only */
+      let m = if not(sys_enable_rvc()) then m else update_C(m, v.C());
+      /* Suppress updates to F/D if they are read-only or would become an illegal combination */
+      if   not(sys_enable_fdext())
+      then m
+      else if (v.D() == 0b1 & v.F() == 0b0)
+      then {
+      	if m.F() == 0b1
+      	then update_D(update_F(m, 0b0), 0b0)
+      	else m
+      }
+      else update_D(update_F(m, v.F()), v.D())
```

Comment:
```suggestion
      if   not(sys_enable_fdext())
      then m
      else update_D(update_F(m, v.F()), v.D() & v.F())
```
is a simpler way to implement the specified semantics I believe. Yours is pretty convoluted and hard to understand.

Reply:
Suggested changes are not according to the spec. Privileged spec says about `misa`:

> If an ISA feature x depends on an ISA feature y, then attempting to enable feature x but disable feature y results in both features being disabled. For example, setting “F”=0 and “D”=1 results in both “F” and “D” being cleared.

Reply:
I disagree; please tell me what specific case it gets wrong.

Reply:
To explain:

If v.F is 0 then F is being disabled or staying disabled, and the line becomes `update_D(update_F(m, 0b0), v.D() & 0b0)` i.e. `update_D(update_F(m, 0b0), 0b0)`.

If v.F is 1 then F is being enabled or staying enabled, and the line becomes `update_D(update_F(m, 0b1), v.D() & 0b1)` i.e. `update_D(update_F(m, 0b1), v.D())`.

This is precisely the behaviour you want.

Reply:
If previously F is enabled(F=1) and D is disabled(D=0). Then, attempting to enable D (D=1) and disable F (F=0) should result in both being disabled(F=0 & D=0) which is according to the spec. But in the suggested changes it will update the value in any case.

Reply:
v.F() is 0 and v.F() & v.D() is 0 & 1 = 0 in that case, so I do not see what the problem is. It does the right thing.

Reply:
Got it. I will update it.